HarmonyCore/FileIO/DataObjectIOBase.dbl

```diff
@@ -1,832 +1,1249 @@
+        ;;; <param name="dataObject">The passed in data object.</param>
+        ;;; <param name="keyOfReference">The required key-of-reference to return the key value for.</param>
+        ;;; <returns></returns>
+        public method GetKeyValue, a
+            in req dataObject, T
+            in req keyOfReference, int
+        proc
+            mreturn GetAlphaFileInfo("KEYVAL", dataObject.SynergyRecord, keyOfReference)
+        endmethod
+        
+        ;;; <summary>
+        ;;;  Read the first record from the file and assign the located record to a $$HarmonyCore Data Object$$.
+        ;;; </summary>
+        ;;; <param name="dataObject">The $$HarmonyCore Data Object$$ to propogate with the located data.</param>
+        ;;; <remarks>
+        ;;; The file will be read on the key index that was set by the last read operation, or the default of primary if no previous operation was performed.  
+        ;;; The record will be locked if the file is opened in update mode and the record is successfully located.
+        ;;; </remarks>
+        public method ReadFirstRecord, FileAccessResults
+            out req dataObject, T
+        proc
+            mreturn ReadFirstRecord(dataObject, false)
+        endmethod
+        
+        ;;; <summary>
+        ;;;  Read the first record from the file and assign the located record to a $$HarmonyCore Data Object$$.
+        ;;; </summary>
+        ;;; <param name="dataObject">The $$HarmonyCore Data Object$$ to propogate with the located data.</param>
+        ;;; <param name="noLock">If passed as TRUE then the record will not be locked, regardless of the file open mode.</param>
+        ;;; <remarks>
+        ;;; The file will be read on the key index that was set by the last read operation, or the default of primary if no previous operation was performed.
+        ;;; The record will not be locked.  To update the record you should use the UpdateRecordUsingGRFA method.
+        ;;; </remarks>
+        public method ReadFirstRecord, FileAccessResults
+            out req dataObject, T
+            in req noLock, boolean
+        proc
+            dataObject = ^null
+
+            if (noLock) then
+            begin
+                data grfaBuffer, a10
+                data result = this.Read(^m(MemoryMapper.fld(1:mDataObjectMetadata.StructureSize), mRecordBuffer), grfaBuffer, 0, "", FindOperation.FromBeginingOfFile, Q_NO_LOCK)
+                if(result == FileAccessResults.Success)
+                    dataObject = mDataObjectMetadata.MakeNew(^m(MemoryMapper.fld(1:mDataObjectMetadata.StructureSize), mRecordBuffer), grfaBuffer)
+                
+                mreturn result
+            end
+            else
+            begin
+                data cancelTokenSource, @CancellationTokenSource, new CancellationTokenSource()
+                data cancelToken, @CancellationToken, cancelTokenSource.Token
+                data dataObj = dataObject
+
+                lambda lamb()
```

Comment:
I think you should use this instead of creating a task and timer
https://www.synergex.com/docs/#lrm/lrmChap4WAIT.htm?Highlight=wait

Reply:
Resolved. See 22e8674720c64fb94bccbeac1df0bb178de513f9